### PR TITLE
Add weight argument to fct_infreq

### LINKED ISSUE #261: `fct_infreq()` should support weights
similarly to `fct_lump()` .

``` r
library(forcats)

x <- letters[c(1:3, 2)]
w <- c(4, 1, 3, 1)

x
#> [1] "a" "b" "c" "b"
w
#> [1] 4 1 3 1

fct_infreq(x)
#> [1] a b c b
#> Levels: b a c

fct_infreq(x, w = w)
#> Error in fct_infreq(x, w = w): unused argument (w = w)

# Should behave like:
fct_reorder(x, w, sum, .desc = TRUE)
#> [1] a b c b
#> Levels: a c b
```

<sup>Created on 2020-04-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by hadley**
Fixes #261

### CODE REVIEW

File: R/reorder.R [hunk 1]
```diff
@@ -134,10 +134,23 @@ fct_inorder <- function(f, ordered = NA) {
 
 #' @export
 #' @rdname fct_inorder
-fct_infreq <- function(f, ordered = NA) {
+#' @param weight Optional vector of numeric values to weight by.
+fct_infreq <- function(f, weight = NULL, ordered = NA) {
```
Comment (hadley) [inline R/reorder.R:138]: Is `weight` the right name?

File: NEWS.md [hunk 2]
```diff
@@ -1,5 +1,8 @@
 # forcats (development version)
 
+* `fct_infreq()` gains the ability to `weight` frequency using a supplemental
+  variable.
```
Comment (DavisVaughan) [inline NEWS.md:4]:
```suggestion
  variable (#261).
```

File: R/reorder.R (hunk 1 above)
Comment (DavisVaughan) [reply R/reorder.R:138]: `fct_lump()` and the other functions in that family use `w` FWIW

I don't think there are any other weighted forcats functions to compare with.

File: R/reorder.R [hunk 3]
```diff
@@ -134,10 +134,23 @@ fct_inorder <- function(f, ordered = NA) {
...
-  lvls_reorder(f, order(table(f), decreasing = TRUE), ordered = ordered)
+  if (is.null(weight)) {
+    weight <- rep(1, length(f))
+  } else {
+    if (!is.numeric(weight)) {
+      cli::cli_abort("{.arg weight} must be a numeric vector, not {.obj_type_friendly weight}.")
+    }
+    if (length(weight) != length(f)) {
+      cli::cli_abort("{.arg f} and {.arg weight} must be the same length.")
+    }
```
Comment (DavisVaughan) [inline R/reorder.R:149]: Do we care if `weight` has missing values?

File: R/reorder.R [hunk 4]
```diff
@@ -134,10 +134,23 @@ fct_inorder <- function(f, ordered = NA) {
...
 #' @export
 #' @rdname fct_inorder
-fct_infreq <- function(f, ordered = NA) {
+#' @param weight Optional vector of numeric values to weight by.
+fct_infreq <- function(f, weight = NULL, ordered = NA) {
   f <- check_factor(f)
 
-  lvls_reorder(f, order(table(f), decreasing = TRUE), ordered = ordered)
+  if (is.null(weight)) {
+    weight <- rep(1, length(f))
```
Comment (DavisVaughan) [inline R/reorder.R:142]:
```suggestion
    weight <- rep(1L, length(f))
```

File: tests/testthat/test-reorder.R [hunk 5]
```diff
@@ -74,6 +74,23 @@ test_that("fct_infreq respects missing values", {
   expect_equal(levels(fct_infreq(f)), c(NA, "b", "a"))
 })
 
+test_that("fct_infreq() respects weights", {
+  x <- c("a", "b", "b", "c")
+  f <- factor(x, exclude = FALSE)
```
Comment (DavisVaughan) [inline tests/testthat/test-reorder.R:79]:
```suggestion
  f <- factor(x)
```

Is there a reason for `exclude`? Or use `fct()`?

File: R/reorder.R [hunk 6]
```diff
@@ -134,10 +134,23 @@ fct_inorder <- function(f, ordered = NA) {
...
+  } else {
+    if (!is.numeric(weight)) {
+      cli::cli_abort("{.arg weight} must be a numeric vector, not {.obj_type_friendly weight}.")
+    }
+    if (length(weight) != length(f)) {
+      cli::cli_abort("{.arg f} and {.arg weight} must be the same length.")
+    }
+  }
```
Comment (DavisVaughan) [inline R/reorder.R:150]: We are ok with fractional values I guess?

File: R/reorder.R (hunk 3 above)
Comment (DavisVaughan) [reply R/reorder.R:149]: I see we have `check_weights()` which could help here. And `check_calc_levels()`

File: R/reorder.R (hunk 1 above)
Comment (DavisVaughan) [inline R/reorder.R:138]: Do you expect the argument ordering here to matter? Probably not much? i.e. if someone specified `ordered` without naming the argument. Seems unlikely

**Review by DavisVaughan — APPROVED** [review]

File: tests/testthat/test-reorder.R (hunk 5 above)
Comment (hadley) [reply tests/testthat/test-reorder.R:79]: Oops

File: R/reorder.R (hunk 6 above)
Comment (hadley) [reply R/reorder.R:150]: Yeah, that seems fine to me.

File: R/reorder.R (hunk 1 above)
Comment (hadley) [reply R/reorder.R]: Yeah, I think that seems ok

**Pull request merged**